### PR TITLE
CPLAT-7606 Improve handling of “repeat” errors thrown from components wrapped by an ErrorBoundary

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
## Motivation
While testing the recent `3.0.0-alpha.0` release in a consumer library, a problem was discovered.

Currently, our `ErrorBoundary` component - when `props.fallbackUIRenderer` is not specified - will attempt to remount its child component tree when an error is caught.  In most cases, this is a sane default - however - if a component in that tree throws __every time it mounts__ - this causes a huge problem, since it creates a sort of "infinite loop" of `component throws in componentDidMount -> ErrorBoundary remounts the componnet -> component throws in componentDidMount -> ...`.  This infinite loop has the potential to completely lock up the main CPU thread.

1. Add some internal component fields to `ErrorBoundaryComponent` that keep track of the error that was thrown, and the "component stack" from the `info` object made available from ReactJS when a component wrapped in an error boundary throws.
2. Use these new fields when a second error is thrown to evaluate whether the second error is identical to the first - and whether it was thrown from the exact same place in the tree.
3. If it is identical - and the errors happen within 5 seconds of each other, render a non-interactive String representation of the DOM that existed before the error was thrown.
4. Allow consumers to get ErrorBoundary out of error states by either passing in a new child, or by calling the new `reset` instance method.
5. Add lots of tests and docs.

#### Release Notes
Improve handling of “repeat” errors thrown from components wrapped by an ErrorBoundary component.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: @greglittlefield-wf @kealjones-wk @joebingham-wk @sydneyjodon-wk 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
The items listed below may not capture all security relevant changes. Before providing a security review, be sure to review the entire PR for security impact.
<details>
<summary>(3) Security relevant changes were detected</summary>
<li>Watched keyword <code>innerHtml</code> in <code>lib/src/component/error_boundary.dart</code> line(s) ['321'] added</li>
<li>Watched keyword <code>InnerHtml</code> in <code>lib/src/component/error_boundary.dart</code> line(s) ['345'] added</li>
<li>Watched keyword <code>dangerouslySetInnerHTML</code> in <code>lib/src/component/error_boundary.dart</code> line(s) ['346'] added</li>
</details>

## Action Items
* Obtain a [security review](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); reviewer should pay special attention to insights listed above
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Comment by maxwellpeterson-wf** [comment]
+1 security
- Our CSP protects us from any potential XSS that this allows.

File: lib/src/component/error_boundary.dart [hunk 1]
```diff
@@ -183,4 +250,136 @@ class ErrorBoundaryComponent<T extends ErrorBoundaryProps, S extends ErrorBounda
...
+      _startIdenticalErrorTimer();
+    }
+  }
+
+  // [2.2]
+  ReactElement _renderStringDomAfterUnrecoverableErrors(_, __) {
+    return (Dom.div()
+      ..key = 'ohnoes'
+      ..addTestId('ErrorBoundary.unrecoverableErrorInnerHtmlContainerNode')
+      ..['dangerouslySetInnerHTML'] = {'__html': _domAtTimeOfError} ?? ''
```
Comment (joebingham-wk) [inline lib/src/component/error_boundary.dart]: What are the cases when the null check will return `''`? Because the `Map` is always set with a key, won't it always be non-null?

File: test/over_react/component/error_boundary_test.dart [hunk 2]
```diff
@@ -47,7 +46,9 @@ void main() {
         document.body.append(mountNode);
         var jacketOfFlawedComponentWithNoErrorBoundary = mount(Flawed()(), mountNode: mountNode);
         expect(mountNode.children, isNotEmpty, reason: 'test setup sanity check');
-        jacketOfFlawedComponentWithNoErrorBoundary.getNode().click();
+        final buttonThatShouldThrowAnErrorWhenClicked = queryByTestId(jacketOfFlawedComponentWithNoErrorBoundary.getInstance(), 'flawedComponent_flawedButton');
+        expect(buttonThatShouldThrowAnErrorWhenClicked, isNotNull, reason: 'test setup sanity check');
+        buttonThatShouldThrowAnErrorWhenClicked.click();
```
Comment (joebingham-wk) [inline test/over_react/component/error_boundary_test.dart]: #nit is there a best practice about when to add space between lines? This is definitely readable without extra lines, but it's also a time where I would personally add a couple to group things and make it a little easier on the eyes.

File: test/over_react/component/error_boundary_test.dart [hunk 3]
```diff
@@ -145,6 +167,446 @@ void main() {
...
+      });
+
+      group('and consecutive errors are thrown from the same component', () {
+        Future<Null> triggerErrorsThatSignifyAnUnrecoverableComponent() async {
+          getFlawedButtonNode().click();
+          expect(calls.length, 1, reason: 'test setup sanity check');
+          expect(calls.single.keys.single, isNot('onComponentIsUnrecoverable'), reason: 'test setup sanity check');
+
+          calls.clear();
+          await new Future.delayed(const Duration(milliseconds: identicalErrorFrequencyToleranceInMs - 50));
```
Comment (greglittlefield-wf) [inline test/over_react/component/error_boundary_test.dart]: #nit magic number `50`.

In tests like these, I like to just do multiples of the main number to avoid having to keep all these in sync:
```suggestion
          await new Future.delayed(identicalErrorFrequencyTolerance ~/ 2);
```

File: test/over_react/component/error_boundary_test.dart [hunk 4]
```diff
@@ -145,6 +167,446 @@ void main() {
...
+        });
+      });
+    });
+
+    group('gracefully handles errors in its tree when `props.fallbackUIRenderer` is not set', () {
+      List<Map<String, List>> calls;
+      DivElement mountNode;
+      var flawedRenderedInstance;
+      var nestedFlawedRenderedInstance;
+      const identicalErrorFrequencyToleranceInMs = 100;
```
Comment (greglittlefield-wf) [inline test/over_react/component/error_boundary_test.dart]: This value should probably be a little bit larger to reduce flakiness on slow machines. Maybe 500ms? Or, you could alternatively add a `retry` to the group.

Also #nit making this a `Duration` could make this easier to work with and reduce the chance of messing up units

File: lib/src/component/error_boundary.dart [hunk 5]
```diff
@@ -183,4 +250,136 @@ class ErrorBoundaryComponent<T extends ErrorBoundaryProps, S extends ErrorBounda
...
+
+  /// Resets the [ErrorBoundary] to a non-error state.
+  ///
+  /// This can be called manually on the component instance using a `ref` -
+  /// or by passing in a new child instance after a child has thrown an error.
+  void reset() {
+    setState((newState()
+      ..hasError = false
+      ..showFallbackUIOnError = props.fallbackUIRenderer != null
+    ), _resetInternalErrorTracking);
```
Comment (greglittlefield-wf) [inline lib/src/component/error_boundary.dart]: Why is this being done after the state change as opposed to before it?

The case that comes to mind is the reset call from `componentDidUpdate`... I feel like we should start with a clean slate whenever the child changes.

File: lib/src/component/error_boundary.dart (hunk 1 above)
Comment (greglittlefield-wf) [reply lib/src/component/error_boundary.dart]: Is this supposed to be on `_domAtTimeOfError`?
```suggestion
      ..['dangerouslySetInnerHTML'] = {'__html': _domAtTimeOfError ?? ''}
```

Maybe we should add a test where the child component is a or renders a fragment, which would represent a case where `_domAtTimeOfError` would be `null`:
```dart
ErrorBoundary([ // Can't use variadic children or the fragment will be interpreted as two separate children
  // A list-style "fragment" as the single child.
  [
    (Dom.div()..key = 1)(),
    (FaultyComponent()..key = 2)(),
  ],
])
```
Or actually, maybe even an empty component would do...

Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Nice catch on this typo @joebingham-wk !!!

File: test/over_react/component/error_boundary_test.dart (hunk 2 above)
Comment (aaronlademann-wf) [reply test/over_react/component/error_boundary_test.dart]: We don't have any hard and fast rules unfortunately. I usually try to group things - but I'll take another pass at this.

File: lib/src/component/error_boundary.dart (hunk 1 above)
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Fixed the typo, added edge case test coverage in https://github.com/Workiva/over_react/pull/350/commits/a9e864eeef2abd045c977aa1377569d7fcb3d399

File: test/over_react/component/error_boundary_test.dart (hunk 2 above)
Comment (aaronlademann-wf) [reply test/over_react/component/error_boundary_test.dart]: Updated in https://github.com/Workiva/over_react/pull/350/commits/320c15f4d808bc07d58455e5c7575082410cf5c5

File: lib/src/component/error_boundary.dart [hunk 6]
```diff
@@ -90,53 +91,110 @@ UiFactory<ErrorBoundaryProps> ErrorBoundary = _$ErrorBoundary;
...
   ///     )(
   ///       // The rest of your component tree
   ///     )
   ///
   /// > See: <https://reactjs.org/docs/react-component.html#componentdidcatch>
+  ///
+  /// > Related: [onComponentIsUnrecoverable]
   _ComponentDidCatchCallback onComponentDidCatch;
 
+  /// An optional callback that will be called _(when [ErrorBoundaryProps.fallbackUIRenderer] is not set)_
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/a78f71fe5cfc650eb5804211b601655b0f1c4727

File: lib/src/component/error_boundary.dart [hunk 7]
```diff
@@ -90,53 +91,110 @@ UiFactory<ErrorBoundaryProps> ErrorBoundary = _$ErrorBoundary;
...
+  ///
+  /// When this callback is called, the tree wrapped by this [ErrorBoundary] has "crashed" - and is completely
+  /// non-functional. Instead of re-mounting the component tree, the [ErrorBoundary] will simply render a
+  /// non-interactive `String` representation of the DOM that was captured at the time of the error.
+  ///
+  /// Once this happens, regaining interactivity within the tree wrapped by this [ErrorBoundary] is possible by:
+  ///
+  /// 1. Passing in a new child
+  ///   _(preferably one that will not repeatedly throw errors when the [ErrorBoundary] mounts it)_.
+  /// 2. Calling the [ErrorBoundaryComponent.reset] instance method using a `ref`.
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/a78f71fe5cfc650eb5804211b601655b0f1c4727

File: lib/src/component/error_boundary.dart [hunk 8]
```diff
@@ -90,53 +91,110 @@ UiFactory<ErrorBoundaryProps> ErrorBoundary = _$ErrorBoundary;
...
-  /// > Default: [ErrorBoundaryComponent._renderDefaultFallbackUI]
+  /// > Related: [onComponentIsUnrecoverable], [onComponentDidCatch]
   _FallbackUiRenderer fallbackUIRenderer;
+
+  /// The amount of time that is "acceptable" between consecutive identical errors thrown from a component
+  /// within the tree wrapped by this [ErrorBoundary].
+  ///
+  /// If [fallbackUIRenderer] is not set, and more than one identical error is thrown consecutively by
+  /// the exact same component anywhere within the tree wrapped by this [ErrorBoundary] - more often than
+  /// the specified duration, the [ErrorBoundary] will:
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/a78f71fe5cfc650eb5804211b601655b0f1c4727

File: lib/src/component/error_boundary.dart [hunk 9]
```diff
@@ -90,53 +91,110 @@ UiFactory<ErrorBoundaryProps> ErrorBoundary = _$ErrorBoundary;
...
+  ///
+  ///   1. Call [onComponentIsUnrecoverable].
+  ///   2. Stop attempting to re-mount the tree (to protect the main thread from being locked up).
+  ///   3. Render a non-interactive `String` representation of the DOM at the time of the error.
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/a78f71fe5cfc650eb5804211b601655b0f1c4727

File: lib/src/component/error_boundary.dart [hunk 10]
```diff
@@ -183,4 +250,136 @@ class ErrorBoundaryComponent<T extends ErrorBoundaryProps, S extends ErrorBounda
...
+  /// * __Before the timer's callback fires__ - internal component logic will treat the second error
+  ///   as an unrecoverable one that has the potential to lock the main thread.
+  /// * __After the timer's callback fires__ - internal component logic will NOT treat the second error
+  ///   as an unrecoverable one.
+  ///
+  /// > Not used when [ErrorBoundaryProps.fallbackUIRenderer] is set.
+  void _startIdenticalErrorTimer() {
+    if (_identicalErrorTimer != null) return;
+
+    _identicalErrorTimer = Timer(props.identicalErrorFrequencyTolerance, _resetInternalErrorTracking);
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/078c84fe7ca0dd4d4053bbb0582e602791a90d8c

File: lib/src/component/error_boundary.dart [hunk 11]
```diff
@@ -183,4 +250,136 @@ class ErrorBoundaryComponent<T extends ErrorBoundaryProps, S extends ErrorBounda
...
+  // ---------------------------------------------- /\ ----------------------------------------------
+
+  String _domAtTimeOfError;
+  /*Error||Exception*/dynamic _lastError;
+  String _lastErrorInfo;
+  Timer _identicalErrorTimer;
+
+  /// Called by [componentDidCatch].
+  void _handleErrorInComponentTree(/*Error||Exception*/dynamic error, /*NativeJavascriptObject*/dynamic jsErrorInfo) {
+    bool sameErrorWasThrownTwiceConsecutively =
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/f6b1f43f9404a97af67a833f8994f6a78585f168

File: test/over_react/component/error_boundary_test.dart [hunk 12]
```diff
@@ -145,6 +167,446 @@ void main() {
...
+
+          test('when reset() is called', () {
+            jacket.getDartInstance().reset();
+            expect(jacket.getDartInstance().state.hasError, isFalse);
+            expect(jacket.getDartInstance().state.showFallbackUIOnError, isTrue);
+            expect(getByTestId(jacket.getInstance(), 'dummyChild'), isNotNull);
+            expect(getByTestId(jacket.getInstance(), 'fallbackNode'), isNull);
+          });
+
+          test('when a new child is passed in', () {
```
[the remark this answers is not in the document]
Comment (aaronlademann-wf) [reply test/over_react/component/error_boundary_test.dart]: Test added via https://github.com/Workiva/over_react/pull/350/commits/a9e864eeef2abd045c977aa1377569d7fcb3d399

File: test/over_react/component/error_boundary_test.dart (hunk 3 above)
Comment (aaronlademann-wf) [reply test/over_react/component/error_boundary_test.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/a9e864eeef2abd045c977aa1377569d7fcb3d399

File: test/over_react/component/error_boundary_test.dart (hunk 4 above)
Comment (aaronlademann-wf) [reply test/over_react/component/error_boundary_test.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/a9e864eeef2abd045c977aa1377569d7fcb3d399

File: lib/src/component/error_boundary.dart (hunk 5 above)
Comment (aaronlademann-wf) [reply lib/src/component/error_boundary.dart]: Addressed in https://github.com/Workiva/over_react/pull/350/commits/1998d8daf855f671b382255e5978a0fb41130428

File: test/over_react/component/error_boundary_test.dart (hunk 2 above)
Comment (joebingham-wk) [reply test/over_react/component/error_boundary_test.dart]: Looks great, thank you for addressing that!

File: test/over_react/component/error_boundary_test.dart [hunk 13]
```diff
@@ -123,13 +133,30 @@ void main() {
...
+          component.setState(component.newState()
+            ..hasError = true
+            ..showFallbackUIOnError = true
+          );
 
-          // Using throws for now since this is temporary, and the throwsUnimplementedError doesn't work here for some reason
-          expect(() => component.setState(component.newState()..hasError = true), throws);
+          expect(getByTestId(jacket.getInstance(), 'dummyChild'), isNull,
+              reason: 'The child component tree should have been removed from the dom');
+          expect(jacket.getNode(), hasAttr(defaultTestIdKey, 'ErrorBoundary.unrecoverableErrorInnerHtmlContainerNode'));
```
Comment (greglittlefield-wf) [inline test/over_react/component/error_boundary_test.dart]: Can we assert here that the static HTML got rendered? I'm not seeing that tested anywhere

Comment (aaronlademann-wf) [reply test/over_react/component/error_boundary_test.dart]: It is tested in the "gracefully handles errors in its tree when `props.fallbackUIRenderer` is not set" section: 

https://github.com/Workiva/over_react/pull/350/files#diff-820fada62e348ea7fccc19d5cc72d0f6R342

Comment (greglittlefield-wf) [reply test/over_react/component/error_boundary_test.dart]: Ah, sorry, I missed that. Thanks!

**Comment by greglittlefield-wf** [comment]
Works well in WVC and over_react example

QA +1

@Workiva/release-management-p

**Comment by maxwellpeterson-wf** [comment]
+1 security refresh
